### PR TITLE
remote_server: Remove dependency on libssl and libcrypto

### LINKED ISSUE #15599: Unable to Execute zed-remote-server-stable Command on Ubuntu Server 22.04 Due to Missing libssl.so.1.1
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

When using SSH to remotely connect to an Ubuntu Server 22.04 development environment, the execution of the zed-remote-server-stable command fails due to a dependency on libssl.so.1.1, which has been deprecated in Ubuntu 22.04.

#### Error Log:
```txt
2024-08-01T10:04:34.823854Z [ERROR] failed to run command: .local/zed-remote-server-stable: /lib/x86_64-linux-gnu/libssl.so.1.1: version `OPENSSL_1_1_1' not found (required by .local/zed-remote-server-stable)
```

### Environment

no

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>

</pre></details>

### PR DESCRIPTION
**Pull request by osiewicz**
Fixes: #15599
Release Notes:

- N/A

### CODE REVIEW

File: crates/collab/src/rpc.rs [hunk 1]
```diff
@@ -992,13 +992,7 @@ impl Server {
...
             tracing::info!("connection opened");
 
-            let http_client = match IsahcHttpClient::new() {
-                Ok(http_client) => Arc::new(http_client),
-                Err(error) => {
-                    tracing::error!(?error, "failed to create HTTP client");
-                    return;
-                }
-            };
+            let http_client = IsahcHttpClient::new(None);
```
Comment (mikayla-maki) [inline crates/collab/src/rpc.rs:995]: We should probably roll back this change. We probably don't want to remove isahc from collab, even if we do switch to `ureq`

File: crates/gpui/src/app.rs [hunk 2]
```diff
@@ -139,6 +139,15 @@ impl App {
         self
     }
 
+    /// Set the http client for the application
+    pub fn with_http_client(self, http_client: impl HttpClient) -> Self {
```
Comment (osiewicz) [inline crates/gpui/src/app.rs:143]: nit: Could we go with `Arc<dyn HttpClient>` as the type of http_client?

File: crates/gpui/src/app.rs [hunk 3]
```diff
@@ -139,6 +139,15 @@ impl App {
...
+        let mut context_lock = self.0.borrow_mut();
+        let http_client = Arc::new(http_client);
+        context_lock.http_client = http_client;
+        drop(context_lock);
```
Comment (osiewicz) [inline crates/gpui/src/app.rs:147]: We don't need to drop the lock explicitly ig

**Comment by osiewicz** [comment]
Hey @MatthiasGrandl @huacnlee, 
This PR changes how gpui app should be initialized; it used to provide a http client by default, but we needed to make that customizable. If your apps depend on e.g. fetching images from Internet, bumping your projects past that commit might break things. On the upside, your apps no longer need to depend on isahc if they don't need to.

I plan to merge this PR tomorrow after the releases are cut.

**Comment by huacnlee** [comment]
👌

**Comment by MatthiasGrandl** [comment]
No worries from my side!

**Comment by huacnlee** [comment]
Hi @osiewicz 

Today, I have upgrade GPUI to latest. And then I realize that we must to `set_http_client`. Because the `img` element need, if it load image from a remote URL, that will need `cx.http_client`.

So, maybe you can consider can we have a http client in GPUI by default, then we do not need set it again.

**Comment by osiewicz** [comment]
This is precisely why we have done it. In Zed we wanted to get rid of a default http client (isahc) and instead let user pick their own.

**Comment by huacnlee** [comment]
I see, thank you.

**Pull request merged**